crawl-ref/source/files.cc

```diff
@@ -2382,6 +2491,69 @@ static FILE* _make_bones_file(string * return_gfilename)
     return nullptr;
 }
 
+#define GHOST_PERMASTORE_SIZE 10
+#define GHOST_PERMASTORE_REPLACE_CHANCE 5
+
+static bool _update_permastore(const vector<ghost_demon> &ghosts, bool wiz_cmd)
+{
+    if (ghosts.empty())
+        return false;
+
+    // this read is not locked...
+    vector<ghost_demon> permastore = _load_permastore_ghosts(wiz_cmd);
+
+    bool rewrite = false;
+    int i = 0;
+    while (permastore.size() < GHOST_PERMASTORE_SIZE && i < ghosts.size())
+    {
+        // TODO: heuristics to make this as distinct as possible; maybe
+        // create a new name?
+        permastore.push_back(ghosts[i]);
+#ifdef DGAMELAUNCH
+        // randomize name for online play
+        permastore.back().name = make_name();
+#endif
+        i++;
+        rewrite = true;
+    }
+    if (i > 0)
+        dprf("Permastoring %d ghosts", i);
+    if (!rewrite && x_chance_in_y(GHOST_PERMASTORE_REPLACE_CHANCE, 100))
+    {
+        int rewrite_i = random2(permastore.size());
+        permastore[rewrite_i] = ghosts[0];
+#ifdef DGAMELAUNCH
+        permastore[rewrite_i].name = make_name();
+#endif
+        rewrite = true;
+    }
+
+    if (rewrite)
+    {
+        string permastore_file = _bones_permastore_file();
+        FILE *gfil = lk_open("wb", permastore_file);
```

Comment:
Surely we can afford at least `gfile`, although personally I'd prefer `ghost_file`.